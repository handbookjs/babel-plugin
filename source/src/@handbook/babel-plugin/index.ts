import { NodePath, Visitor } from '@babel/traverse';
import types from '@babel/types';
import fs from 'fs';
import nodePath from 'path';

const PACKAGE_NAME = '@handbook/source';
const NAMESPACE = 'namespace';
const SOURCE = 'source';

interface Babel {
  types: typeof types;
}

interface BabelPlugin {
  visitor: Visitor<{
    opts?: {
      target?: string;
      runtime?: string;
    };
    file: {
      path: NodePath;
    };
  }>;
}

interface BabelOptions {
  cwd: string;
  filename: string;
}

interface PluginOptions {
  sourceRoot?: string;
  targetExtensions?: string[];
}

function getExtension(
  loc: string,
  ...extensions: string[]
): string | undefined {
  return (
    process.env.HANDBOOK_TEST_EXT ??
    extensions.find((ext) => fs.existsSync(loc + ext))
  );
}

function getAbsoluteFileLocation(
  importPath: string,
  babelOptions: BabelOptions,
  pluginOptions: PluginOptions,
): string | undefined {
  if (/^\./.test(importPath)) {
    return nodePath.resolve(
      nodePath.dirname(babelOptions.filename),
      importPath,
    );
  }

  const dir: string = nodePath.dirname(importPath);

  const sourceRoot: string =
    pluginOptions.sourceRoot ?? nodePath.resolve(babelOptions.cwd, 'src');

  if (fs.existsSync(nodePath.resolve(sourceRoot, dir))) {
    return nodePath.resolve(sourceRoot, importPath);
  } else if (fs.existsSync(nodePath.resolve(babelOptions.cwd, dir))) {
    return nodePath.resolve(babelOptions.cwd, importPath);
  } else {
    return undefined;
  }
}

function getSourceRoot(
  babelOptions: BabelOptions,
  pluginOptions: PluginOptions,
): string {
  if (typeof pluginOptions.sourceRoot === 'string')
    return pluginOptions.sourceRoot;

  if (fs.existsSync(nodePath.resolve(babelOptions.cwd, 'src'))) {
    return nodePath.resolve(babelOptions.cwd, 'src');
  }

  return babelOptions.cwd;
}

function getImportPath<T>(
  t: typeof types,
  args0:
    | types.Expression
    | types.SpreadElement
    | types.JSXNamespacedName
    | types.ArgumentPlaceholder,
): string | undefined {
  if (t.isStringLiteral(args0)) {
    return args0.value;
  } else if (
    t.isCallExpression(args0) &&
    t.isIdentifier(args0.callee) &&
    args0.callee.name === 'require' &&
    args0.arguments.length > 0 &&
    t.isStringLiteral(args0.arguments[0])
  ) {
    return args0.arguments[0].value;
  } else if (
    t.isArrowFunctionExpression(args0) &&
    t.isCallExpression(args0.body) &&
    t.isImport(args0.body.callee) &&
    args0.body.arguments.length > 0 &&
    t.isStringLiteral(args0.body.arguments[0])
  ) {
    return args0.body.arguments[0].value;
  }

  return undefined;
}

export function transformSourceCallExpression(
  t: typeof types,
  path: NodePath<types.CallExpression>,
  babelOptions: BabelOptions,
  pluginOptions: PluginOptions,
) {
  if (path.node.arguments.length < 1) return;

  const args0 = path.node.arguments[0];

  const importPath: string | undefined = getImportPath(t, args0);

  if (!importPath) return;

  const absoulteFileLocation: string | undefined = getAbsoluteFileLocation(
    importPath,
    babelOptions,
    pluginOptions,
  );

  if (!absoulteFileLocation) return;

  const targetExtensions: string[] = pluginOptions.targetExtensions ?? [
    '.tsx',
    '.jsx',
    '.js',
    '.ts',
    '.mjs',
    '.mdx',
    '.css',
    '.sass',
    '.scss',
    '.less',
    '.svg',
    '.yml',
    '.yaml',
    '.json',
  ];

  const pathHasExtname: boolean = fs.existsSync(absoulteFileLocation);

  const ext: string | undefined = pathHasExtname
    ? nodePath.extname(absoulteFileLocation)
    : getExtension(absoulteFileLocation, ...targetExtensions);

  if (!ext) return;

  const sourceRoot: string = getSourceRoot(babelOptions, pluginOptions);

  path.node.arguments[0] = t.objectExpression([
    t.objectProperty(t.identifier('module'), args0 as types.Expression),
    t.objectProperty(
      t.identifier('source'),
      t.memberExpression(
        t.callExpression(t.identifier('require'), [
          t.stringLiteral(`!!raw-loader!${importPath}`),
        ]),
        t.identifier('default'),
      ),
    ),
    t.objectProperty(
      t.identifier('filename'),
      t.stringLiteral(
        nodePath
          .relative(sourceRoot, absoulteFileLocation)
          .replace(/\\/g, '/') + (pathHasExtname ? '' : ext),
      ),
    ),
  ]);
}

//eslint-disable-next-line import/no-anonymous-default-export
export default function (
  { types: t }: Babel,
  pluginOptions: PluginOptions,
): BabelPlugin {
  let babelOptions: BabelOptions | null = null;
  const localNames: Map<typeof NAMESPACE | typeof SOURCE, string> = new Map();

  return {
    visitor: {
      Program(path: NodePath<types.Program>) {
        babelOptions = path.hub.file.opts;
        localNames.clear();
      },

      ImportDeclaration(path: NodePath<types.ImportDeclaration>) {
        if (path.node.source.value !== PACKAGE_NAME) return;

        for (const specifier of path.node.specifiers) {
          // import * as ns from '@handbook/source'
          // import ns from '@handbook/source'
          if (
            t.isImportNamespaceSpecifier(specifier) ||
            t.isImportDefaultSpecifier(specifier)
          ) {
            localNames.set(NAMESPACE, specifier.local.name);
          }
          // import { source } from '@handbook/source'
          // import { source as another } from '@handbook/source'
          else if (t.isImportSpecifier(specifier)) {
            switch (specifier.imported.name) {
              case SOURCE:
                localNames.set(specifier.imported.name, specifier.local.name);
                break;
            }
          }
        }
      },
      CallExpression(path: NodePath<types.CallExpression>) {
        if (!babelOptions) return;

        function hasCall(
          path: NodePath<types.CallExpression>,
          callName: typeof SOURCE,
        ): boolean {
          return (
            // import { source } from '@handbook/source'
            // import { source as another } from '@handbook/source'
            (localNames.has(callName) &&
              t.isIdentifier(path.node.callee, {
                name: localNames.get(callName),
              })) ||
            // import * as ns from '@handbook/source'
            // import ns from '@handbook/source'
            (localNames.has(NAMESPACE) &&
              'object' in path.node.callee &&
              'property' in path.node.callee &&
              t.isIdentifier(path.node.callee.object, {
                name: localNames.get(NAMESPACE),
              }) &&
              t.isIdentifier(path.node.callee.property, { name: callName }))
          );
        }

        if (hasCall(path, SOURCE)) {
          transformSourceCallExpression(t, path, babelOptions, pluginOptions);
        }
      },
    },
  };
}
