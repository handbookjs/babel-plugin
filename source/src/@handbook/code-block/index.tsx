export * from './components/MDXCodeBlock';
export * from './components/CodeBlock';
