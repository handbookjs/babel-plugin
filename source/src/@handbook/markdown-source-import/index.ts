import { glob } from '@ssen/promised';
import * as fs from 'fs-extra';
import flatten from 'lodash.flattendeep';
import path from 'path';
import prettier from 'prettier';
import git from 'simple-git';
import unified from 'unified';

export interface MarkdownSourceImportParams {
  cwd?: string;
  filePatterns: string[];
  ignore?: string[];
  gitAdd?: boolean;
}

export const processor = unified()
  .use(require('remark-parse'))
  .use(require('@handbook/remark-split-html-lines'))
  .use(require('@handbook/remark-magic-comments'))
  .use(require('remark-stringify'), {
    listItemIndent: 1,
  });

export async function markdownSourceImport({
  cwd = process.cwd(),
  filePatterns,
  ignore = ['**/node_modules/**', './node_modules/**'],
  gitAdd = false,
}: MarkdownSourceImportParams) {
  const files: string[] = await Promise.all(
    filePatterns.map((pattern) =>
      glob(`${cwd}/${pattern}`, {
        ignore,
      }),
    ),
  ).then((...group) => flatten(group));

  for (const file of files) {
    const input: string = await fs.readFile(file, 'utf8');

    const { contents } = await processor.process({
      contents: input,
      path: file,
      basename: path.basename(file),
      extname: path.extname(file),
      dirname: path.dirname(file),
      cwd,
    });

    if (typeof contents === 'string') {
      const config: prettier.Options | null = await prettier.resolveConfig(
        file,
        { useCache: true },
      );
      const next: string = prettier.format(contents, {
        singleQuote: true,
        trailingComma: 'all',
        ...config,
        parser: 'markdown',
      });

      if (next !== input && next.length > 0) {
        await fs.writeFile(file, next, { encoding: 'utf8' });

        try {
          if (gitAdd) {
            await git(cwd).add(file);
          }
        } catch (error) {
          console.error(error);
        }
      }
    }
  }
}
