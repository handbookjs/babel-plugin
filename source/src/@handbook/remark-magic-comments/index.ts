import { Plugin } from 'unified';
import { Node } from 'unist';
import { processCommand } from './processCommand';
import { transformCommand } from './transformCommand';
import { isRootNode } from './types';

const plugin: Plugin<[]> = () => (node, { dirname = process.cwd() }): Promise<Node> | void => {
  if (isRootNode(node)) {
    return processCommand(transformCommand(node), dirname);
  }
};

export = plugin;
