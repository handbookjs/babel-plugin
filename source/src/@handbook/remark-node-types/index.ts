export type RootNode = Node & { children: Node[] };
export type HtmlNode = Node & { value: string };

export function isRootNode(node: Node): node is RootNode {
  return node.type === 'root';
}

export function isHtmlNode(node: Node): node is HtmlNode {
  return node.type === 'html' && typeof node.value === 'string';
}

// ---------------------------------------------
// copy from @types/unist
// ---------------------------------------------
/**
 * Syntactic units in unist syntax trees are called nodes.
 */
export interface Node {
  /**
   * The variant of a node.
   */
  type: string;

  /**
   * Information from the ecosystem.
   */
  data?: Data;

  /**
   * Location of a node in a source document.
   * Must not be present if a node is generated.
   */
  position?: Position;

  [key: string]: unknown;
}

/**
 * Information associated by the ecosystem with the node.
 * Space is guaranteed to never be specified by unist or specifications
 * implementing unist.
 */
export interface Data {
  [key: string]: unknown;
}

/**
 * Location of a node in a source file.
 */
export interface Position {
  /**
   * Place of the first character of the parsed source region.
   */
  start: Point;

  /**
   * Place of the first character after the parsed source region.
   */
  end: Point;

  /**
   * Start column at each index (plus start line) in the source region,
   * for elements that span multiple lines.
   */
  indent?: number[];
}

/**
 * One place in a source file.
 */
export interface Point {
  /**
   * Line in a source file (1-indexed integer).
   */
  line: number;

  /**
   * Column in a source file (1-indexed integer).
   */
  column: number;
  /**
   * Character in a source file (0-indexed integer).
   */
  offset?: number;
}
