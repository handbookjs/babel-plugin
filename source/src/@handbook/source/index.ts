export interface SourceModule<T> {
  module: T;
  source: string;
  filename: string;
}

export function source<T>(module: T | SourceModule<T>): SourceModule<T> {
  if ('module' in module && 'source' in module && 'filename' in module) {
    return module as SourceModule<T>;
  }
  throw new Error(
    `Can't find the module. You have to install @handbook/babel-plugin`,
  );
}
