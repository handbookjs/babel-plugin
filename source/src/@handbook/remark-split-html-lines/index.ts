import { isRootNode, Node } from '@handbook/remark-node-types';
import { Plugin } from 'unified';
import { transform } from './transform';

const plugin: Plugin<[]> = () => (node): Promise<Node> | void => {
  if (isRootNode(node)) {
    return transform(node);
  }
};

export = plugin;
