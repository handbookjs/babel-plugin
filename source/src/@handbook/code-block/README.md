# `@handbook/code-block`

[![NPM](https://img.shields.io/npm/v/@handbook/code-block.svg)](https://www.npmjs.com/package/@handbook/code-block)
[![TEST](https://github.com/rocket-hangar/handbook/workflows/Test/badge.svg)](https://github.com/rocket-hangar/handbook/actions?query=workflow%3ATest)
[![codecov](https://codecov.io/gh/rocket-hangar/handbook/branch/master/graph/badge.svg)](https://codecov.io/gh/rocket-hangar/handbook)

## CodeBlock React components

<a href="https://rocket-handbook-example.netlify.app" target="_blank">

  <img src="https://raw.githubusercontent.com/rocket-hangar/handbook/master/doc-assets/code-block.png" width="800" style="max-width: 800px" />

</a>

Use React component.

```jsx
import { CodeBlock } from "@handbook/code-block";

function Component(sourceCode: string) {
  return <CodeBlock language="js">{sourceCode}</CodeBlock>;
}
```

Set default code block for mdx documents.

```jsx
import { MDXCodeBlock } from "@handbook/code-block";

const components = {
  pre: (props) => <div {...props} />,
  code: MDXCodeBlock,
};
```

## API

<!-- source components/CodeBlock.tsx --pick "CodeBlockProps CodeBlock" -->

[components/CodeBlock.tsx](components/CodeBlock.tsx)

```tsx
export function CodeBlock({ children, language, theme }: CodeBlockProps) {}

export interface CodeBlockProps {
  /**
   * your code block theme
   *
   * you can choose one of dracula, duotoneDark, duotoneLight, github, nightOwl, nightOwlLight, oceanicNext, palenight, shadesOfPurple, synthwave84, ultramin and vsDark
   *
   * @example import github from 'prism-react-renderer/themes/github'
   *
   * @see <rootDir>/node_modules/prism-react-renderer/themes
   */
  theme?: PrismTheme;
  /** source code */
  children: string;
  /** language */
  language: Language;
}
```

<!-- /source -->

<!-- source components/MDXCodeBlock.tsx --pick "MDXCodeBlockProps MDXCodeBlock" -->

[components/MDXCodeBlock.tsx](components/MDXCodeBlock.tsx)

```tsx
export function MDXCodeBlock({
  theme = vsDark,
  children,
  className = "language-javascript",
}: MDXCodeBlockProps) {}

export interface MDXCodeBlockProps {
  /**
   * your code block theme
   *
   * you can choose one of dracula, duotoneDark, duotoneLight, github, nightOwl, nightOwlLight, oceanicNext, palenight, shadesOfPurple, synthwave84, ultramin and vsDark
   *
   * @example import github from 'prism-react-renderer/themes/github'
   *
   * @see <rootDir>/node_modules/prism-react-renderer/themes
   */
  theme?: PrismTheme;
  /** source string */
  children: string;
  /** css className */
  className?: string;
}
```

<!-- /source -->

## How to get source codes

These code block components print source codes only.

If you want to get source codes into your Web App use [`@handbook/source`](https://www.npmjs.com/package/@handbook/source), [`@handbook/babel-plugin`](https://www.npmjs.com/package/@handbook/babel-plugin) and [`@handbook/typescript-source-sampler`](https://www.npmjs.com/package/@handbook/typescript-source-sampler).

You can print your source code on your Web App easier.

```tsx
import React from "react";
import { render } from "react-dom";
import { source } from "@handbook/source";
import { sampling } from "@handbook/typescript-source-sampler";
import { CodeBlock } from "@handbook/code-block";

const module = source(require("./source/hello"));
const samples = sampling({ source: module.source, samples: ["Class", "func"] });

function App() {
  return <CodeBlock language="typescript" children={samples.get("Class")} />;
}

render(<App />, document.querySelector("#app"));
```

## See more

- [`@handbook/*`](https://github.com/rocket-hangar/handbook) This package is one of `@handbook/*` packages. Go to the project home and see more details.
