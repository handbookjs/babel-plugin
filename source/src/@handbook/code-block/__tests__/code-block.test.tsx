import { CodeBlock, MDXCodeBlock } from '@handbook/code-block';
import { render } from '@testing-library/react';
import prettier from 'prettier';
import * as React from 'react';

function format(code: string): string {
  return prettier.format(code, { parser: 'typescript' });
}

const source = format(`
interface Test {
  s: string;
  a: number;
  b: number;
}
`);

describe('@handbook/code-block', () => {
  test('should render <MDXCodeBlock>', async () => {
    // Arrange
    const { container } = render(<MDXCodeBlock className="language-typescript" children={source} />);

    // Assert
    expect(container.querySelector('.token.keyword')?.textContent).toBe('interface');
    expect(container.querySelector('.token.class-name')?.textContent).toBe('Test');
    expect(format(container.textContent ?? '')).toBe(source);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('should render <CodeBlock>', async () => {
    // Arrange
    const { container } = render(<CodeBlock language="typescript" children={source} />);

    // Assert
    expect(container.querySelector('.token.keyword')?.textContent).toBe('interface');
    expect(container.querySelector('.token.class-name')?.textContent).toBe('Test');
    expect(format(container.textContent ?? '')).toBe(source);
    expect(container.innerHTML).toMatchSnapshot();
  });
});
