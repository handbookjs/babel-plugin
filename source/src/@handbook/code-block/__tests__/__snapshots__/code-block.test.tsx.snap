// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@handbook/code-block should render <CodeBlock> 1`] = `"<pre class=\\"prism-code language-typescript\\" style=\\"color: rgb(156, 220, 254); background-color: rgb(30, 30, 30); padding: 1em;\\"><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token keyword\\" style=\\"color: rgb(86, 156, 214);\\">interface</span><span class=\\"token plain\\"> </span><span class=\\"token class-name\\" style=\\"color: rgb(78, 201, 176);\\">Test</span><span class=\\"token plain\\"> </span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">{</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\">  s</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">:</span><span class=\\"token plain\\"> </span><span class=\\"token builtin\\" style=\\"color: rgb(86, 156, 214);\\">string</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">;</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\">  a</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">:</span><span class=\\"token plain\\"> </span><span class=\\"token builtin\\" style=\\"color: rgb(86, 156, 214);\\">number</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">;</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\">  b</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">:</span><span class=\\"token plain\\"> </span><span class=\\"token builtin\\" style=\\"color: rgb(86, 156, 214);\\">number</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">;</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\"></span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">}</span></div></pre>"`;

exports[`@handbook/code-block should render <MDXCodeBlock> 1`] = `"<pre class=\\"prism-code language-typescript\\" style=\\"color: rgb(156, 220, 254); background-color: rgb(30, 30, 30); padding: 1em;\\"><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token keyword\\" style=\\"color: rgb(86, 156, 214);\\">interface</span><span class=\\"token plain\\"> </span><span class=\\"token class-name\\" style=\\"color: rgb(78, 201, 176);\\">Test</span><span class=\\"token plain\\"> </span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">{</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\">  s</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">:</span><span class=\\"token plain\\"> </span><span class=\\"token builtin\\" style=\\"color: rgb(86, 156, 214);\\">string</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">;</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\">  a</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">:</span><span class=\\"token plain\\"> </span><span class=\\"token builtin\\" style=\\"color: rgb(86, 156, 214);\\">number</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">;</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\">  b</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">:</span><span class=\\"token plain\\"> </span><span class=\\"token builtin\\" style=\\"color: rgb(86, 156, 214);\\">number</span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">;</span><span class=\\"token plain\\"></span></div><div class=\\"token-line\\" style=\\"color: rgb(156, 220, 254);\\"><span class=\\"token plain\\"></span><span class=\\"token punctuation\\" style=\\"color: rgb(212, 212, 212);\\">}</span></div></pre>"`;
